### DISCUSSION TITLE
Error EINVAL when binding XDP to Ethernet on aws EC2 ARM

### DISCUSSION THREAD

**Discussion by wujunzhuo** [cilium/ebpf#783] [Q&A]
The xdp example runs normally for `lo`(Local Loopback); however, it returns error when switched to ethernet. Are there any special requirements for ARM servers? Since this issue doesn't happen on my x86 servers.

```sh
ubuntu@ip-172-31-34-8:~/ebpf/examples/xdp$ sudo ./xdp ens5
2022/09/02 07:47:34 could not attach XDP program: create link: invalid argument
```

```sh
ubuntu@ip-172-31-34-8:~/ebpf/examples/xdp$ ifconfig 
ens5: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 3840
        inet 172.31.34.8  netmask 255.255.240.0  broadcast 172.31.47.255
        inet6 fe80::c79:d1ff:fe3a:9a96  prefixlen 64  scopeid 0x20<link>
        ether 0e:79:d1:3a:9a:96  txqueuelen 1000  (Ethernet)
        RX packets 346816  bytes 479539290 (479.5 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 92136  bytes 8270897 (8.2 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 408  bytes 44919 (44.9 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 408  bytes 44919 (44.9 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

```sh
ubuntu@ip-172-31-34-8:~/ebpf/examples/xdp$ go version
go version go1.18.5 linux/arm64

ubuntu@ip-172-31-34-8:~/ebpf/examples/xdp$ uname -a
Linux ip-172-31-34-8 5.15.0-1019-aws #23-Ubuntu SMP Wed Aug 17 18:35:04 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux

ubuntu@ip-172-31-34-8:~/ebpf/examples/xdp$ cat /etc/issue
Ubuntu 22.04.1 LTS \n \l
```

**Comment by lmb** [comment]
IIRC XDP doesn't work with jumbo frames, the MTU on ens5 is likely too large. Try reducing it to something like 2000 and try again.

**Reply by wujunzhuo** [reply]
> IIRC XDP doesn't work with jumbo frames, the MTU on ens5 is likely too large. Try reducing it to something like 2000 and try again.

Thnaks @lmb 

Yes, the MTU matters. On x86 servers, 3840 is the largest possible size (the default value is 9001 on EC2s). But tuning the MTU seems has no effect on this ARM server: I have tried to reduce it to lower than 800, but it still doesn't work.

**Reply by diskcat** [reply]
I have the same question, I try change MTU size but it doesn't work. I want to know how to deal this question?

**Reply by mejedi** [reply]
@diskcat you also need to lower the number of combined channels to 1/2 of the maximum, see https://trying2adult.com/what-is-xdp-and-how-do-you-use-it-in-linux-amazon-ec2-example/ `4. THE NUMBER OF CHANNELS AVAILABLE FOR THE RX/TX QUEUES MUST BE AT LEAST HALF OF THE MAXIMUM AVAILABLE CHANNELS ON THE NIC.`

**Reply by lmb** [reply]
@mejedi want to post that as a top level reply so i can mark it as the answer?
